mm2src/mm2_core/src/sql_connection_pool.rs

```diff
@@ -0,0 +1,332 @@
+use crate::mm_ctx::{log_sqlite_file_open_attempt, path_to_dbdir, MmCtx};
+use async_std::sync::RwLock as AsyncRwLock;
+use common::log::error;
+use db_common::async_sql_conn::AsyncConnection;
+use db_common::sqlite::rusqlite::Connection;
+use futures::channel::mpsc::{channel, Receiver, Sender};
+use futures::lock::Mutex as AsyncMutex;
+use gstuff::try_s;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex, MutexGuard, RwLock};
+
+pub const ASYNC_SQLITE_DB_ID: &str = "KOMODEFI.db";
+const SYNC_SQLITE_DB_ID: &str = "MM2.db";
+const SQLITE_SHARED_DB_ID: &str = "MM2-shared.db";
+
+/// Represents the kind of database connection ID: either shared or single-db.
+enum DbIdConnKind {
+    Shared,
+    Single,
+}
+
+/// A pool for managing SQLite connections, where each connection is keyed by a unique string identifier.
+#[derive(Clone)]
+pub struct SqliteConnPool {
+    connections: Arc<RwLock<HashMap<String, Arc<Mutex<Connection>>>>>,
+    // default db_id
+    rmd160_hex: String,
+    // default shared_db_id
+    shared_db_id: String,
+    db_root: Option<String>,
+}
+
+impl SqliteConnPool {
+    /// Initializes a single-user database connection.
+    pub fn init(ctx: &MmCtx, db_id: Option<&str>) -> Result<(), String> {
+        Self::init_impl(ctx, db_id, DbIdConnKind::Single)
+    }
+
+    /// Initializes a shared database connection.
+    pub fn init_shared(ctx: &MmCtx, db_id: Option<&str>) -> Result<(), String> {
+        Self::init_impl(ctx, db_id, DbIdConnKind::Shared)
+    }
+
+    /// Internal implementation to initialize a database connection.
+    fn init_impl(ctx: &MmCtx, db_id: Option<&str>, kind: DbIdConnKind) -> Result<(), String> {
```

Comment:
Suggestion:
if first create SqlliteConnPool instance I guess you may reuse the existing sqlite_conn_impl(), something like that:
```
    fn init_impl(ctx: &MmCtx, db_id: Option<&str>, kind: DbIdConnKind) -> Result<(), String> {

        // Create a new connection pool if it does not exist
        if ctx.sqlite_conn_pool.as_option().is_none() {
            let db_root = ctx.conf["dbdir"].as_str();
            try_s!(ctx.sqlite_conn_pool.pin(Self {
                connections: Arc::new(RwLock::new(HashMap::new())),
                rmd160_hex: ctx.rmd160_hex(),
                shared_db_id: hex::encode(*ctx.shared_db_id()),
                db_root: db_root.map(|d| d.to_owned())
            }));
        }

        if let Some(conn_pool) = ctx.sqlite_conn_pool.as_option() {
            conn_pool.sqlite_conn_impl(db_id, kind);
        }
        Ok(())
    }
```

Reply:
I guess this is not needed since this cause an unnecessary check in `sqlite_conn_impl()` which we can can avoid by just writing directly from here.

https://github.com/KomodoPlatform/komodo-defi-framework/blob/b19d2ba222898a80a8777a690bd0944dc358f531/mm2src/mm2_core/src/sql_connection_pool.rs#L118-L123